atomic_reactor/utils/resource_manager.py

```diff
@@ -270,45 +561,64 @@ def from_config(cls, config: dict):
           enabled: true
           auth: qa-vm-secret-filepath
           username: cloud-user
+          slots: 3
           ...
         hostname-remote-host2:
           ...
         """
-
-        # Add enabled clusters to list hosts
-        hosts = [RemoteHost(hostname=hostname, username=attr['username'], ssh_keyfile=attr['auth'])
-                 for hostname, attr in config.items() if attr['enabled']]
+        hosts = []
+        for hostname, attr in config.items():
+            if not attr.get("enabled", False):
+                continue
+            host = RemoteHost(
+                hostname=hostname, username=attr["username"], ssh_keyfile=attr["auth"],
+                slots=attr.get("slots", 1)
+            )
+            # Check whether host is operational before use it
+            if host.is_operational:
+                hosts.append(host)
 
         return cls(hosts)
 
-    @backoff.on_exception(
-        backoff.constant,
-        LockRetry,
-        max_time=TIMEOUT_FAIL_BUILD,
-        jitter=None,
-    )
-    def lock_resource(self, prid: str) -> LockedResource:
-        """ Lock resources for build """
+    def lock_resource(self, prid: str) -> Optional[LockedResource]:
+        """
+        Lock resource for a pipelinerun
+
+        :param prid: str, pipelinerun ID
+        """
+
+        if not self.hosts:
+            logger.error("This is no available remote host in pool")
+            return None
 
-        # Slots should be randomized here to avoid always locking the lowest numbers first
         resources = []
         for host in self.hosts:
             available_slots = host.available_slots()
-            random.shuffle(available_slots)
+            if not available_slots:
+                logger.info("%s: no available slots", host.hostname)
+                continue
+            logger.info("%s: available slots: %s", host.hostname, available_slots)
             resources.append((host, available_slots))
 
-        # Sort list based on number of available slots
-        resources.sort(key=lambda x: len(x[1]), reverse=True)
+        if not resources:
+            logger.error("There is no remote host slot available for pipelinerun %s", prid)
+            return None
+
+        # Sort list based on ratio of available_slots/all_slots
+        resources.sort(key=lambda x: len(x[1])/x[0].slots, reverse=True)
 
-        # Try to lock resources
+        # Try to lock a remote host slot for pipelinerun
         for host, slots in resources:
             for slot in slots:
```

Comment:
Do we still want try the slots in a random order?

Reply:
It's unnecessary, there is no difference among slots on the same host except the number, so the try order doesn't matter.

Reply:
ah, I see, I didn't realized this is what randomizing trying to resolve, changed in updated version.